process/headerCheck/headerVersioningHandler.go

```diff
@@ -0,0 +1,167 @@
+package headerCheck
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"sort"
+
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/process"
+)
+
+const wildcard = "*"
+
+type headerVersioningHandler struct {
+	referenceChainID []byte
+	versions         []config.VersionByEpochs
+	defaultVersion   string
+}
+
+// NewHeaderVersioningHandler returns a new instance of a structure capable of handling the versions of a header
+// based on the provided epoch
+func NewHeaderVersioningHandler(
+	referenceChainID []byte,
+	versionsByEpochs []config.VersionByEpochs,
+	defaultVersion string,
+) (*headerVersioningHandler, error) {
+
+	if len(referenceChainID) == 0 {
+		return nil, ErrInvalidReferenceChainID
+	}
+
+	hvh := &headerVersioningHandler{
+		referenceChainID: referenceChainID,
+		defaultVersion:   defaultVersion,
+	}
+	var err error
+	hvh.versions, err = hvh.prepareVersions(versionsByEpochs)
+	if err != nil {
+		return nil, err
+	}
+
+	err = hvh.checkVersionLength([]byte(defaultVersion))
+	if err != nil {
+		return nil, err
+	}
+
+	return hvh, err
+}
+
+func (hvh *headerVersioningHandler) prepareVersions(versionsByEpochs []config.VersionByEpochs) ([]config.VersionByEpochs, error) {
+	sort.Slice(versionsByEpochs, func(i, j int) bool {
+		return versionsByEpochs[i].StartEpoch < versionsByEpochs[j].StartEpoch
+	})
+
+	currentEpoch := uint32(0)
+	for _, ver := range versionsByEpochs {
+		if ver.StartEpoch >= ver.EndEpoch {
+			return nil, fmt.Errorf("%w, StartEpoch is greater or equal to EndEpoch, version %s",
+				ErrInvalidVersionOnEpochValues, ver.Version)
+		}
+		if currentEpoch != ver.StartEpoch {
+			return nil, fmt.Errorf("%w for version %s",
+				ErrInvalidVersionFoundGaps, ver.Version)
+		}
+		if len(ver.Version) > core.MaxSoftwareVersionLengthInBytes {
+			return nil, fmt.Errorf("%w for version %s",
+				ErrInvalidVersionStringTooLong, ver.Version)
+		}
+
+		currentEpoch = ver.EndEpoch
+	}
+
+	return versionsByEpochs, nil
+}
+
+// GetVersion returns the version by providing the epoch
+func (hvh *headerVersioningHandler) GetVersion(epoch uint32) string {
+	ver := hvh.getMatchingVersion(epoch)
+	if ver == wildcard {
+		return hvh.defaultVersion
+	}
+
+	return ver
+}
+
+// GetVersion returns the version by providing the epoch
+func (hvh *headerVersioningHandler) getMatchingVersion(epoch uint32) string {
+	//TODO add a small cache here?
```

Comment:
this TODO does not say enough

Reply:
The question was if we should add a small sized cache that will improve the responsiveness of this function since this will be called for each and every header received. Since epochs change with a low frequency, caching will remove the iteration of the provided config array.

Reply:
And yes, I want to remove the TODO before merging this PR :)

Reply:
removed TODO, added cache

Reply:
+1